### DISCUSSION TITLE
Which Rules are Inlined?

### DISCUSSION THREAD

**Discussion by berchn** [tree-sitter/tree-sitter#955] [Q&A] [answered]
In "The Grammar DSL" section of the documentation, it specifies `inline` as a public field which can be set to modify the behavior of the generated parser. This presumably contains all of rules that should be inlined; however, any rule that starts with an underscore appears to be inlined in practice. What is the intended behavior?

**Comment by sogaiu** [comment]
This may not be an answer to your question, but FWIW, the behavior with respect to underscore has some docs here: https://tree-sitter.github.io/tree-sitter/creating-parsers#hiding-rules

**Comment by maxbrunsfeld** [comment, marked as answer]
The hidden rules generally aren't inlined (except for in limited cases where the generator can prove that it's behavior-preserving). They are just hidden from the syntax tree to make the tree nicer to use.

In terms of user-facing behavior, the specification is just that any rule in `inline` is inlined.

**Reply by berchn** [reply]
As an example, the `import_statement` rule for Python contains `_import_list`.
`_import_list` further contains a `"FIELD"` named `"name"`. `_import_list` does not
appear in the `"inline"` rules set.

```Python
import a,b;

"""
module [0, 0] - [1, 0])
  import_statement [0, 0] - [0, 10])
    name: dotted_name [0, 7] - [0, 8])
      identifier [0, 7] - [0, 8])
    name: dotted_name [0, 9] - [0, 10])
      identifier [0, 9] - [0, 10])
"""
```
The `"name"` fields appear under `import_statement` instead of being hidden from
the tree, hence the appearance of being inlined.

An example of behavior I would expected for a hidden rule would be `_semicolon`
in Python where its presence is completely hidden from the output--albeit it doesn't have a subtree.

Is there a way to determine which of these rules behave this way without looking up the hidden rule and checking for potential children?

**Reply by maxbrunsfeld** [reply]
Yeah, this is the intended behavior of hiding a rule; it’s children are not all recursively hidden. That wouldn’t be very useful IMO.

Inlining is different than this, because it actually changes the parsing process, affects where ambiguities can or can’t occur, etc.
